packages/core/src/util/applicationUnderMonitoring.js

```diff
@@ -113,18 +113,44 @@ function getMainPackageJsonPathStartingAtDirectory(startDirectory, cb) {
 
   if (!startDirectory) {
     // No explicit starting directory for searching for the main package.json has been provided, use the Node.js
-    // process' main module as the starting point.
-    const mainModule = process.mainModule;
+    // "require.main" module as the starting point.
+
+    // NOTE: "require.main" is undefined when the Instana collector is required inside a
+    // preloaded module using "--require". "process.mainModule" is not, but it is deprecated and should
+    // no longer been used.
+    let mainModule = require.main;
 
     if (!mainModule) {
       // This happens
       // a) when the Node CLI is evaluating an expression, or
-      // b) when the REPL is used, or
-      // c) when we have been pre-required with the --require/-r command line flag
-      // In particular for case (c) we want to try again later. This is handled in the individual metrics that rely on
-      // evaluating the package.json file.
-      return process.nextTick(cb);
+      // b) when the REPL is used
+      // c) when we have been pre-required with the --require/-r command line flag.
+      //
+      // In particular for case (c) we can try again later and wait for the main module to be loaded.
+      // But usually that is not necessary because the initialisation of the collector takes longer than
+      // Node.js having not loaded the main module. Still, it is "ok" to keep the retry mechanismn inside
+      // the individual metrics (e.g. name.js) just for safety.
+      //
+      // But when the application is using ES modules and they require the Instana collector
+      // with "--require file.cjs", neither `require.main` nor the deprecated `process.mainModule`
+      // will have a value, because ES modules use `import.meta` and soon `import.main`.
+      // See https://github.com/nodejs/modules/issues/274
+
+      // eslint-disable-next-line max-len
+      // See https://github.com/nodejs/node/blob/472edc775d683aed2d9ed39ca7cf381f3e7e3ce2/lib/internal/modules/run_main.js#L79
+      // Node.js is using `process.argv[1]` internally as main file path.
+      // Check whether a module was preloaded and use process.argv[1] as filename.
+      // @ts-ignore
+      if (process._preload_modules && process._preload_modules.length > 0) {
```

Comment:
The fix starts here. I wasn't able to think of a different solution. But satisfied because Node.js is doing the same internally.

Reply:
One alternative I could think of is to walk through the module cache until you find a module that has import.meta.main === true. But I think that would be much worse than your current solution. 👍   Plus, the property is not even implemented yet.